contracts/oracle/DynamicVoltSystemOracle.sol

```diff
@@ -0,0 +1,216 @@
+        uint256 periodPriceChange = cachedOraclePrice * actualChangeRate;
+        uint256 priceDelta = periodPriceChange * timeDelta / TIMEFRAME / 1e18;
+
+        return cachedOraclePrice + priceDelta;
+    }
+
+    /// ------------- Governor Only API -------------
+
+    /// @notice Set the base VOLT rate.
+    /// This will also refresh the actual VOLT rate.
+    /// Callable only by governor.
+    /// @param newBaseChangeRate new value for the baseChangeRate.
+    function updateBaseRate(uint256 newBaseChangeRate) external onlyGovernor {
+        /// first, compound interest
+        _compoundInterest();
+
+        uint256 oldBaseChangeRate = baseChangeRate; // SLOAD
+        baseChangeRate = newBaseChangeRate; // SSTORE
+        emit BaseRateUpdated(
+            block.timestamp,
+            oldBaseChangeRate,
+            newBaseChangeRate
+        );
+
+        /// if too few liquid reserves, adjust rate up
+        uint256 newActualChangeRate = DynamicVoltRateModel(rateModel).getRate(
+            newBaseChangeRate,
+            PCVOracle(pcvOracle).lastLiquidVenuePercentage()
+        );
+        uint256 oldActualChangeRate = actualChangeRate; // SLOAD
+        actualChangeRate = newActualChangeRate; // SSTORE
+        emit ActualRateUpdated(
+            block.timestamp,
+            oldActualChangeRate,
+            newActualChangeRate
+        );
+    }
+
+    /// @notice Set the reference to the Rate Model.
+    /// Callable only by governor.
+    /// @param newRateModel address of the new rate model.
+    function setRateModel(address newRateModel) external onlyGovernor {
+        address oldRateModel = rateModel; // SLOAD
+        rateModel = newRateModel; // SSTORE
+        // emit event
+        emit RateModelUpdated(block.timestamp, oldRateModel, newRateModel);
+    }
+
+    /// ------------- PCV Oracle Only API -------------
+
+    /// @notice Only callable by the PCV Oracle, updates the actual rate.
+    /// @param liquidPercentage the percentage of PCV that is liquid,
+    /// expressed with 18 decimals.
+    /// @dev this function does not need the `isGlobalReentrancyLocked` modifier,
+    /// because it is only called from the PCVOracle, either through the functions
+    /// `updateLiquidBalance` or `updateIlliquidBalance`, that have the modifier,
+    /// or from `addVenues` that is a governor-only action that will only execute
+    /// during DAO proposals.
+    function updateActualRate(uint256 liquidPercentage) external {
+        require(msg.sender == pcvOracle, "MGO: Not PCV Oracle");
+
+        /// first, compound interest
+        _compoundInterest();
+
+        /// if too few liquid reserves, adjust rate up
+        uint256 newActualChangeRate = DynamicVoltRateModel(rateModel).getRate(
+            baseChangeRate,
+            liquidPercentage
+        );
+        uint256 oldActualChangeRate = actualChangeRate; // SLOAD
+        actualChangeRate = newActualChangeRate; // SSTORE
```

Comment:
if there is no delta, we shouldn't write to storage at all or compound interest to save on gas. the reason for this is anytime a user swaps with the PSM, it is going to have to go through and hit this callback. the less writes to storage, the better to cut down on gas costs.

Then we should expose a `compoundInterest` API that allows us to permissionlessly compound the interest only after the time period has passed in case the rate never changed.

Reply:
I removed the public `compoundInterest` function, because it would only be usable after 1 year without any system interaction, we discussed it here : https://github.com/volt-protocol/volt-protocol-core/pull/147#discussion_r1024274530

adding a check to see if `oldActualChangeRate` != `newActualChangeRate` before doing SSTORE